Add support for blob rewrite

mziccard opened the pull request:
This follows from #291. Main changes are:
- Add `rewrite` method to `StorageRpc` and `DefaultStorageRpc`
- Add `BlobRewriter` class to wrap multiple rewrite requests
- Add `rewriter` method to `Storage` and `StorageImpl` that returns a `BlobRewriter`
- Add unit and integration tests

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -865,4 +1016,23 @@ public static Builder builder() {
+   * @throws StorageException upon failure
+   * @see <a href="https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite">Rewrite</a>
+   */
+  BlobRewriter rewriter(RewriteRequest rewriteRequest);
```
Should we call it `rewrite`?

I wonder if we really need to have both copy and rewrite methods?

One option could be to make the copy do the rewrite, so basically change copy to return a `CopyResponse` (similar
to the `BlobRewriter`) with `isDone`, `nextChunk` and `blobInfo`.

We could also make `blobInfo` a blocking call until complete (or have a more explicit method for it).

/cc @BrandonY @mfschwartz

aozarov replied:
If we keep both than we should probably add it `Blob` as well.

mziccard replied:
We could keep the both and document that users should use `rewrite` when copy involves different locations or different storage classes.

For what concerns the naming if we keep them both I'll change `rewriter` to `rewrite` and `BlobRewriter` to `RewriteResponse` it you agree.

Capstan replied:
I'd much prefer you call this low-level wrapping of Rewrite the same name as the API method it's replacing. If you want to synthesize a higher-level Copy that uses Rewrite under the covers but doesn't expose the rewrite token to the client, that seems entirely reasonable. Since under the covers, Copy uses Rewrite and simply fails if Rewrite cannot complete in one request, uniformly using Rewrite seems like the right thing to do until such time as our service offers a service-side alternative.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/spi/DefaultStorageRpc.java:
```diff
@@ -521,4 +522,31 @@ public String open(StorageObject object, Map<Option, ?> options)
+          .rewrite(source.getBucket(), source.getName(), target.getBucket(), target.getName(),
+              target)
+          .setRewriteToken(token)
+          .setMaxBytesRewrittenPerCall(maxByteRewrittenPerCall)
```
I assume `null` is fine?

Also, since it seems that this value ["must be an integral multiple of 1 MiB (1048576)"](https://developers.google.com/resources/api-libraries/documentation/storage/v1/java/latest/com/google/api/services/storage/Storage.Objects.Rewrite.html#getMaxBytesRewrittenPerCall()) I think it would be nicer the API accepts it in MB (and the rpc layer will multiple the value by 1048576).

mziccard replied:
Yes null sets no value.
And I agree we can make the API accept MB and do the multiplication in the RPC layer.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java:
```diff
@@ -174,4 +175,8 @@ StorageObject copy(StorageObject source, Map<Option, ?> sourceOptions,
   void write(String uploadId, byte[] toWrite, int toWriteOffset, StorageObject dest,
       long destOffset, int length, boolean last) throws StorageException;
+
+  RewriteResponse rewrite(StorageObject source, Map<Option, ?> sourceOptions,
```
One problem with this is that maxByteRewrittenPerCal must be the same value between calls.

We could change the RPC layer to deal with it with an open/continue model where open returns an opaque
value that includes all the information it needs for subsequent calls or we could just document it.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java:
```diff
@@ -0,0 +1,203 @@
+/**
+ * Google Storage blob rewriter.
+ */
+public final class BlobRewriter {
```
We could make it `Restorable` and support re-write continuation between processes (similar to a write channel), no?

mziccard replied:
Sure thing!

mziccard replied:
I'll wait until #296 gets merged to push this and all other fixes as some changes in #296 affect this PR.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  /**
+   * Size of the blob being copied. Returns {@code null} until the first copy request returns.
```
Did you consider to return `BlobRewriter` after sending the first RPC?

In that case `blobSize` will always have a value, `rewriteToken` will be `null` only when `isDone` is `true`
and any permission issues will be discovered upon the rewrite request.

mziccard replied:
I wrote it like this at first. With the first call to RPC in `writer`. I did not like that for rewrites that require only one RPC the `BlobRewriter.copyChunk` method was basically never used/unuseful. But I don't mind changing this.

aozarov replied:
That (sending first RPC upon `rewrite` call) would be my preference.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java:
```diff
@@ -0,0 +1,203 @@
+  /**
+   * Returns the token to be used to rewrite the next chunk of the blob.
+   */
+  public String rewriteToken() {
```
Is there any use for it to the user in our API?

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java:
```diff
@@ -0,0 +1,203 @@
+  /**
+   * Returns the number of bytes written. 
+   */
+  public BigInteger totalBytesRewritten() {
```
Oh, I wonder if we should make it simpler and use long which will support a file of the size more than a [exbibyte](https://en.wikipedia.org/wiki/Exbibyte), especially consider that now max upload size is [5 TB](https://cloud.google.com/storage/docs/json_api/v1/objects/insert). 

Otherwise, we should create a bug and also update BlobInfo.size to use `BigInteger` instead of long.

/cc @BrandonY

mziccard replied:
We already made this assumption in `BlobInfo` right? As of now I think it should be safe to use `Long`.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java:
```diff
@@ -0,0 +1,203 @@
+        isDone = response.getDone();
+        blobSize = response.getObjectSize();
+        totalBytesRewritten = response.getTotalBytesRewritten();
+        target = response.getResource() != null ? BlobInfo.fromPb(response.getResource()) : target;
```
Nit, I find a regular if-condition to be more readable in *this* case

```java
if (response.getResource() != null) {
  target = BlobInfo.fromPb(response.getResource());
}
```

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -596,6 +596,157 @@ public static Builder builder() {
+       *
+       * @return the builder.
+       */
+      public Builder maxBytesRewrittenPerCall(Long maxBytesRewrittenPerCall) {
```
See my comment above, suggesting that we should set it in MB.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -865,4 +1016,23 @@ public static Builder builder() {
   List<Boolean> delete(BlobId... blobIds);
+
+  /**
+   * Returns a {@link BlobRewriter} object for the provided {@code rewriteRequest}. If source and
```
{@code RewriteRequest}

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -865,4 +1016,23 @@ public static Builder builder() {
+   *   }}
+   * </pre>
+   *
+   * @throws StorageException upon failure
```
I don't think it does now, but that would not be the case if we always send the first RPC in this request as I suggested above.

mziccard commented:
@aozarov I addressed all comments: 
- split `StorageRpc.rewrite` into `StorageRpc.openRewrite` and `StorageRpc.openRewrite`
- use `StorageRpc.openRewrite` and `StorageRpc.openRewrite` to implement copy in `Storage` and `Blob`
- remane `BlobRewriter` to `CopyWriter`
- update and add tests
- update example

I also noticed that `rewrite` (in contrast with `copy`) fails if no content type is specified for the destination object.
```java
storage.copy(CopyRequest.of("bucket", "n1", BlobInfo.builder("bucket", "n2").build()));
```
fails with:
> {
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "required",
    "message": "You must specify the content type of the destination object."
   }
  ],
  "code": 400,
  "message": "You must specify the content type of the destination object."
 }
}

Maybe @Capstan can confirm this?
I updated `StorageExample` to take a content type parameter for `cp`.

aozarov commented:
@mziccard I tried to reproduce the `content-type` requirement using the [API Explorer](https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite?hl=en) but for me it worked fine without supplying it.

mziccard commented:
@aozarov My bad, I was providing a storage object in the request body even though it had no `contentType`, forgot to add the check. I'm fixing this.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/spi/StorageRpc.java:
```diff
@@ -132,6 +133,89 @@ public BatchResponse(Map<StorageObject, Tuple<Boolean, StorageException>> delete
+    public final String rewriteToken;
+    public final Long totalBytesRewritten;
+
+    public RewriteResponse(RewriteRequest rewriteRequest, StorageObject result, Long blobSize,
```
Any reason `blobSize`, `isDone` and `totalBytesRewritten` are not primitives? Can they be `null` after the first call?

mziccard replied:
No good reason

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java:
```diff
@@ -213,19 +213,20 @@ public Blob update(BlobInfo blobInfo, BlobTargetOption... options) {
    */
-  public Blob copyTo(BlobId targetBlob, BlobSourceOption... options) {
+  public CopyWriter copyTo(BlobId targetBlob, BlobSourceOption... options) {
     BlobInfo updatedInfo = info.toBuilder().blobId(targetBlob).build();
```
I wonder if that is necessary, rather than using `BlobInfo.builder(targetBlob)` when we are not overriding the metadata?

mziccard replied:
Yes should be better so transfer less data

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,264 @@
+ * Google Storage blob copy writer. This class holds the result of a copy request.
+ * If source and destination blobs do not share the same location or storage class more than one
+ * RPC request is needed to copy the blob. When this is the case {@link #copyChunk()} can be used
+ * to copy to destination other chunks of the source blob.
```
I wonder if something like the following will be slightly more clear:
..RPC request is needed to copy the blob otherwise one or more {@link #copyChunk()} calls are necessary to complete the copy.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,264 @@
+   * Returns {@code null} otherwise.
+   */
+  public BlobInfo result() {
+    return rewriteResponse.result != null ? BlobInfo.fromPb(rewriteResponse.result) : null;
```
Did you consider making this call complete the copy/rewrite operation (and make sure that the method could potentially block)?

mziccard replied:
Yes but I did not like it. If we provide the explicit way to complete the copy I don't think this method should block. Would you prefer it that way?

aozarov replied:
Personally I like the, potentially but (In my assumption) not commonly, blocking result as we do provide a user an option to avoid it if desired by checking with `isDone()` (and I think we should document it).

For most cases (where it would not matter or they don't care) one could just do
`BlobInfo copy = blob.copyTo(BlobId.of(bucket, name)).result();`

I consider the need for, non-subsequent `copyChunk` calls to be the least common usage pattern and therefore I don't think we need to "optimize" for it.
Also, this way `result()` will never be `null`.

However, if you feel more comfortable with a separate, more explicit, call then I am fine with that
too. e.g. `BlobInfo copy = blob.copyTo(BlobId.of(bucket, name)).complete()` or `copyChunks()` which will return the BlobInfo...

What do you think?

mziccard replied:
I see your point on making easier for a user to get the result and avoid returning `null`. I was just in doubt about providing both an explicit  way to do requests (`copyChunk`) and an implicit one(`result()`). But I guess we can suggest the usage of `result()` and let `copyRequest()` to advanced users.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,264 @@
+  /**
+   * Size of the blob being copied.
+   */
+  public Long blobSize() {
```
could this (and bellow) now be primitive?

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -507,42 +508,93 @@ public static Builder builder() {
 
+      /**
+       * Sets the maximum number of megabytes to copy for each RPC call. This parameter is ignored
+       * if source and target blob share the same location and storage class as rewrite is made with
```
We should probably avoid using the "rewrite" term in the user-api (and use "copy") instead.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -507,42 +508,93 @@ public static Builder builder() {
+      }
+
+      /**
+       * Creates a {@code RewriteRequest}.
```
{@code CopyRequest}.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -507,42 +508,93 @@ public static Builder builder() {
+       *
+       * @return the builder.
+       */
+      public Builder megabytesRewrittenPerCall(Long megabytesRewrittenPerCall) {
```
I wonder if the name should be consistent with chunk? 
`megabytesRewrittenPerChunk` and `copyChunk`

mziccard replied:
Yes sounds better

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -755,12 +829,26 @@ public static Builder builder() {
+   * <pre>    {@code CopyWriter copyWriter = service.copy(copyRequest);}
+   *    {@code while(!copyWriter.isDone()) {
+   *       copyWriter.copyChunk();
+   *   }}
```
Do you think we should include the BlobInfo blobInfo = copyWriter.result() in the code sample?

Also, why do you use 2 different {@code } tags?

If you think that result should block (or that we should provide a more explicit method for it) than maybe it is worth showing it as well...

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/StorageImpl.java:
```diff
@@ -444,21 +445,22 @@ public StorageObject call() {
-        public StorageObject call() {
-          return storageRpc.copy(source, sourceOptions, target, targetOptions);
+        public RewriteResponse call() {
+          return storageRpc.openRewrite(new StorageRpc.RewriteRequest(source, sourceOptions, target,
```
We can now remove the `copy` method from `StorageRpc`, right?

aozarov commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java:
```diff
@@ -317,16 +317,15 @@ public void testComposeBlobFail() {
-    assertArrayEquals(BLOB_BYTE_CONTENT, readBytes);
+    BlobId source = BlobId.of(BUCKET, sourceBlobName);
+    assertNotNull(storage.create(BlobInfo.builder(source).build(), BLOB_BYTE_CONTENT));
+    String targetBlobName = "test-rewrite-blob-target";
```
"test-copy-blob-target"

aozarov commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java:
```diff
@@ -317,16 +317,15 @@ public void testComposeBlobFail() {
+    String targetBlobName = "test-rewrite-blob-target";
+    BlobInfo target = BlobInfo.builder(BUCKET, targetBlobName).contentType(CONTENT_TYPE).build();
+    Storage.CopyRequest req = Storage.CopyRequest.of(source, target);
+    CopyWriter rewriter = storage.copy(req);
```
copyWriter

mziccard commented:
I fixed the comments but the one on documenting/throwing on `contentType == null` when setting `CopyRequest.target`. We could open a separate issue for that.

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,269 @@
+
+/**
+ * Google Storage blob copy writer. This class holds the result of a copy request.
+ * If source and destination blobs do not share the same location or storage class more than one
```
Sorry, my fault but I think now we have a strange sentence.

Consider this :
```
If source and destination blobs share the same location and storage class the copy is
completed in one RPC call otherwise one or more {@link #copyChunk} calls are
necessary to complete the copy.
```

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  /**
+   * Returns the updated information for the just written blob. This method might block and issue
```
Would be nice to indicate when it will block.
```
Returns the updated information for the written blob.
Calling this method when {@code isDone()} is {@code false} will 
block until all pending chunks are copied.
```

aozarov replied:
And maybe even plainly saying that this in effect will perform the following:
```
{@code
  while (!isDone()) {
   copyChunk();
  }
}
```

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,269 @@
+   * @throws StorageException upon failure
+   */
+  public BlobInfo result() {
+    while(!isDone()) {
```
```while (!isDone())``` missing a space after the *while*

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  /**
+   * Size of the blob being copied.
```
Nit, "Returns the size of the blob being copied."

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java:
```diff
@@ -0,0 +1,269 @@
+
+  @Override
+  public RestorableState<CopyWriter> capture() {
+    return StateImpl.builder(
```
Fine as is, but I am not sure the Builder in this case giving us much as oppose to just
passing `this` to `StateImpl` constructor.

aozarov commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java:
```diff
@@ -317,16 +317,14 @@ public void testComposeBlobFail() {
-    assertEquals(targetBlobName, remoteBlob.name());
-    byte[] readBytes = storage.readAllBytes(BUCKET, targetBlobName);
-    assertArrayEquals(BLOB_BYTE_CONTENT, readBytes);
+    BlobInfo target = BlobInfo.builder(BUCKET, targetBlobName).contentType(CONTENT_TYPE).build();
```
Maybe in this one we should not override the content-type (and keep the next one to test the override)?

aozarov commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java:
```diff
@@ -334,30 +332,32 @@ public void testCopyBlob() {
+        .build();
+    Storage.CopyRequest req = Storage.CopyRequest.of(source, target);
+    CopyWriter copyWriter = storage.copy(req);
+    assertEquals(copyWriter.result(), storage.get(BUCKET, targetBlobName));
```
Shall we also verify the content-type and metadata that they are what we expect them to be?

aozarov commented on gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java:
```diff
@@ -317,16 +317,14 @@ public void testComposeBlobFail() {
+    BlobInfo target = BlobInfo.builder(BUCKET, targetBlobName).contentType(CONTENT_TYPE).build();
+    Storage.CopyRequest req = Storage.CopyRequest.of(source, target);
+    CopyWriter copyWriter = storage.copy(req);
+    assertEquals(copyWriter.result(), storage.get(BUCKET, targetBlobName));
```
she will verify that metadata from source were copied?

aozarov commented on gcloud-java-storage/src/main/java/com/google/gcloud/storage/Storage.java:
```diff
@@ -755,12 +829,23 @@ public static Builder builder() {
+   * might issue multiple RPC calls depending on blob's size.
+   * <p>
+   * Example usage of copy:
+   * <pre>    {@code BlobInfo blob = service.copy(copyRequest).result();}
```
I think this is great but I also think we should say a bit more about the option of using `isDone`
and `copyChunk` (could be here or in the `CopyWriter` class or maybe my suggestion for the `CopyWriter.response` javadoc suggestion is be sufficient?).

aozarov commented:
Looks great, few more minor comments.

mziccard commented:
Last comments addressed, thanks @aozarov for the thorough review!
